File: modules/crypto-core/src/main/java/com/shterneregen/securelan/crypto/service/impl/DefaultKeyGenerationService.java
package com.shterneregen.securelan.crypto.service.impl;

import com.shterneregen.securelan.crypto.service.KeyGenerationService;

import javax.crypto.KeyGenerator;
import javax.crypto.SecretKey;
import java.security.GeneralSecurityException;
import java.security.KeyPair;
import java.security.KeyPairGenerator;

public final class DefaultKeyGenerationService implements KeyGenerationService {
    private static final int AES_KEY_SIZE = 256;
    private static final int RSA_KEY_SIZE = 2048;

    @Override
    public SecretKey generateAesKey() {
        try {
            KeyGenerator keyGenerator = KeyGenerator.getInstance("AES");
            keyGenerator.init(AES_KEY_SIZE);
            return keyGenerator.generateKey();
        } catch (GeneralSecurityException exception) {
            throw new IllegalStateException("Failed to generate AES key", exception);
        }
    }

    @Override
    public KeyPair generateRsaKeyPair() {
        try {
            KeyPairGenerator keyPairGenerator = KeyPairGenerator.getInstance("RSA");
            keyPairGenerator.initialize(RSA_KEY_SIZE);
            return keyPairGenerator.generateKeyPair();
        } catch (GeneralSecurityException exception) {
            throw new IllegalStateException("Failed to generate RSA key pair", exception);
        }
    }
}

File: modules/crypto-core/src/main/kotlin/com/shterneregen/securelan/crypto/service/impl/DefaultKeyGenerationService.kt
package com.shterneregen.securelan.crypto.service.impl

import com.shterneregen.securelan.crypto.service.KeyGenerationService
import java.security.GeneralSecurityException
import java.security.KeyPair
import java.security.KeyPairGenerator
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey

class DefaultKeyGenerationService : KeyGenerationService {
    override fun generateAesKey(): SecretKey = try {
        KeyGenerator.getInstance("AES").apply { init(AES_KEY_SIZE) }.generateKey()
    } catch (exception: GeneralSecurityException) {
        throw IllegalStateException("Failed to generate AES key", exception)
    }

    override fun generateRsaKeyPair(): KeyPair = try {
        KeyPairGenerator.getInstance("RSA").apply { initialize(RSA_KEY_SIZE) }.generateKeyPair()
    } catch (exception: GeneralSecurityException) {
        throw IllegalStateException("Failed to generate RSA key pair", exception)
    }

    companion object {
        private const val AES_KEY_SIZE = 256
        private const val RSA_KEY_SIZE = 2048
    }
}
